src/main.rs

```diff
@@ -17,45 +17,100 @@
+use astarte_message_hub::AstarteMessageHub;
 
 #[tokio::main]
 async fn main() -> Result<(), AstarteMessageHubError> {
-    use astarte_device_sdk::AstarteDeviceSdk;
-
     env_logger::init();
 
-    //TODO add MessageHubServer and add AstarteHandler::new() on top of AstarteSDK
-    let options = MessageHubOptions::get().await?;
-    let astarte_options = astarte_map_options(&options).await;
-    let _astarte_sdk = AstarteDeviceSdk::new(&astarte_options).await?;
+    // Get the message hub configuration options
+    let msg_hub_opts = MessageHubOptions::get().await?;
+
+    // Initailize an Astarte device
+    let device_sdk = initialize_astarte_device_sdk(msg_hub_opts).await?;
+    info!("Connection to Astarte established.");
+
+    // Create a new Astarte handler
+    let handler = AstarteHandler::new(device_sdk);
+
+    // Create a new message hub
+    let message_hub = AstarteMessageHub::new(handler.clone());
+
+    // Run the protobuf server
+    let addr = "[::1]:50051".parse().unwrap();
+    tonic::transport::Server::builder()
+        .add_service(MessageHubServer::new(message_hub))
+        .serve(addr)
+        .await?;
 
     Ok(())
 }
 
-pub async fn astarte_map_options(
-    opts: &MessageHubOptions,
-) -> astarte_device_sdk::options::AstarteOptions {
-    use astarte_device_sdk::options::AstarteOptions;
-    use astarte_device_sdk::registration;
-
-    let credentials_secret = match &opts.credentials_secret {
-        None => registration::register_device(
-            opts.pairing_token.as_ref().unwrap(),
-            opts.pairing_url.as_ref().unwrap(),
-            opts.realm.as_ref().unwrap(),
-            opts.device_id.as_ref().unwrap(),
+async fn initialize_astarte_device_sdk(
+    mut msg_hub_opts: MessageHubOptions,
+) -> Result<astarte_device_sdk::AstarteDeviceSdk, AstarteMessageHubError> {
+    // If no credential secret is present, register a new device using the Astarte device SDK
+    if msg_hub_opts.credentials_secret.is_none() {
+        msg_hub_opts.credentials_secret = astarte_device_sdk::registration::register_device(
+            msg_hub_opts.pairing_token.as_ref().ok_or_else(|| {
+                AstarteMessageHubError::FatalError(
+                    "Missing pairing token for Astarte device SDK.".to_string(),
+                )
+            })?,
+            msg_hub_opts.pairing_url.as_ref().ok_or_else(|| {
+                AstarteMessageHubError::FatalError(
+                    "Missing pairing URL for Astarte device SDK.".to_string(),
+                )
+            })?,
+            msg_hub_opts.realm.as_ref().ok_or_else(|| {
+                AstarteMessageHubError::FatalError(
+                    "Missing realm for Astarte device SDK.".to_string(),
+                )
+            })?,
+            msg_hub_opts.device_id.as_ref().ok_or_else(|| {
+                AstarteMessageHubError::FatalError(
+                    "Missing device ID for Astarte device SDK.".to_string(),
+                )
```

Comment:
This check is already done, with lower verbosity, by `MessageHubOptions::is_valid` ([ref](https://github.com/astarte-platform/astarte-message-hub/blob/245c252a268887e2c00d6238df68956eb451b4c1/src/config/mod.rs#L82)). I suggest adding all error messages in `is_valid` and replace this block and the next one with a more readable check on `msg_hub_opts.is_valid()` on top of this function

Reply:
I see `is_valid()` depends on further changes in #81. We have two possible choices: revert these two changes or move the ones on the `MessageHubOptions` struct here. It is more efficient to address a similar change only once.

Reply:
Since #81 has to make changes to the configuration while this PR only uses it.
Thinking about it, I think the order of those PRs is flipped.
I think #81 should be merged before this one since it implements changes on which this one is reliant.